compiler/annotation/write_type_check.rs

```diff
@@ -0,0 +1,308 @@
+) -> Option<BTreeSet<(Type, Type)>> {
+    if let Some(first) = constraint_annotations.next() {
+        let mut combinations = first
+            .iter()
+            .flat_map(|(left, right_set)| right_set._iter().map(|right| (left.clone(), right.clone())))
+            .collect::<BTreeSet<_>>();
+        while let Some(annotations) = constraint_annotations.next() {
+            combinations.retain(|(left, right)| {
+                annotations.get(left).map(|right_set| right_set._contains(right)).unwrap_or(false)
+            });
+        }
+        Some(combinations)
+    } else {
+        None
+    }
+}
+
+fn pairs_from_vertex_annotations(
+    input_annotations_variables: &BTreeMap<Variable, Arc<BTreeSet<Type>>>,
+    left: &Vertex<Variable>,
+    right: &Vertex<Variable>,
+    source_span: Option<Span>,
+) -> Result<BTreeSet<(Type, Type)>, TypeInferenceError> {
+    let left_types = input_annotations_variables.get(left.as_variable().as_ref().unwrap()).ok_or(
+        TypeInferenceError::AnnotationsUnavailableForVariableInWrite {
+            variable: right.as_variable().unwrap(),
+            source_span,
+        },
+    )?;
+    let right_types = input_annotations_variables.get(right.as_variable().as_ref().unwrap()).ok_or(
+        TypeInferenceError::AnnotationsUnavailableForVariableInWrite {
+            variable: right.as_variable().unwrap(),
+            source_span,
+        },
+    )?;
+    Ok(left_types.iter().cloned().cartesian_product(right_types.iter().cloned()).collect())
+}
+
+fn check_insert_types_against_match_types<T: ContainsAndIterOnType>(
+    valid_insert_pairs: Arc<BTreeMap<Type, T>>,
+    match_pairs: BTreeSet<(Type, Type)>,
+    constraint: &(impl Into<Constraint<Variable>> + Clone),
+    snapshot: &impl ReadableSnapshot,
+    type_manager: &TypeManager,
+    source_span: Option<Span>,
+) -> Result<(), TypeInferenceError> {
+    let mut invalid_iter = match_pairs.iter().filter(|(left_type, right_type)| {
+        !valid_insert_pairs
+            .get(left_type)
+            .map(|valid_right_types| valid_right_types._contains(right_type))
+            .unwrap_or(false)
+    });
+    if let Some((left_type, right_type)) = invalid_iter.next() {
+        Err(TypeInferenceError::IllegalTypeCombinationForWrite {
+            constraint_name: Into::<Constraint<Variable>>::into(constraint.clone()).name().to_string(),
+            left_type: left_type
+                .get_label(snapshot, type_manager)
+                .map_err(|err| TypeInferenceError::ConceptRead { typedb_source: err })?
+                .scoped_name()
+                .as_str()
+                .to_string(),
+            right_type: right_type
+                .get_label(snapshot, type_manager)
+                .map_err(|err| TypeInferenceError::ConceptRead { typedb_source: err })?
+                .scoped_name()
+                .as_str()
+                .to_string(),
+            source_span,
+        })
+    } else {
+        Ok(())
+    }
+}
+
```

Comment:
`ContainsAndIterOnType` helps template the functions, because `LinksAnnotations` are `BTreeMap<Type, BTreeSet<Type>>` whereas `LeftRightAnnotations` are `BTreeMap<Type, Vec<Type>>`